File: typescript-sdk/src/mcp.ts
import fs from "node:fs";
import path from "node:path";
import { z } from "zod";

const ServerSchema = z.object({
  transport: z.enum(["stdio", "http"]).default("stdio"),
  command: z.string().optional(),
  args: z.array(z.string()).default([]),
  env: z.record(z.string()).default({}),
  enabled: z.boolean().default(true),
  timeout_ms: z.number().int().positive().default(30_000)
});

const McpSchema = z.object({
  version: z.string().default("1"),
  servers: z.record(ServerSchema).default({}),
  tool_policies: z.object({
    allow: z.array(z.string()).default(["*"]),
    deny: z.array(z.string()).default([])
  }).default({ allow: ["*"], deny: [] }),
  defaults: z.object({
    max_tool_calls_per_turn: z.number().int().positive().default(20)
  }).default({ max_tool_calls_per_turn: 20 })
});

export type McpConfig = z.infer<typeof McpSchema>;

export interface CustomTool {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
  execute: (args: Record<string, unknown>) => Promise<unknown> | unknown;
}

export class McpRegistry {
  private readonly tools = new Map<string, CustomTool>();
  register(tool: CustomTool): void {
    this.tools.set(tool.name, tool);
  }
  unregister(name: string): void {
    this.tools.delete(name);
  }
  list(): CustomTool[] {
    return [...this.tools.values()];
  }
}

export class McpConfigLoader {
  loadFromDefaultPath(cwd = process.cwd()): McpConfig {
    const p = path.join(cwd, ".dispersl", "mcp.json");
    if (!fs.existsSync(p)) return McpSchema.parse({});
    const raw = JSON.parse(fs.readFileSync(p, "utf-8")) as Record<string, unknown>;
    return this.interpolateEnv(raw);
  }

  merge(base: McpConfig, override?: Partial<McpConfig>): McpConfig {
    if (!override) return base;
    return McpSchema.parse({
      ...base,
      ...override,
      servers: { ...base.servers, ...(override.servers ?? {}) }
    });
  }

  private interpolateEnv(raw: Record<string, unknown>): McpConfig {
    const text = JSON.stringify(raw).replace(/\$\{([A-Z0-9_]+)\}/g, (_m, name: string) => process.env[name] ?? "");
    return McpSchema.parse(JSON.parse(text));
  }
}

File: python-sdk/src/dispersl/mcp.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable


@dataclass
class MCPTool:
    name: str
    description: str
    parameters: dict[str, Any]
    execute: Callable[[dict[str, Any]], Any]


@dataclass
class MCPConfig:
    version: str = "1"
    servers: dict[str, dict[str, Any]] = field(default_factory=dict)
    tool_policies: dict[str, list[str]] = field(default_factory=lambda: {"allow": ["*"], "deny": []})
    defaults: dict[str, Any] = field(default_factory=lambda: {"max_tool_calls_per_turn": 20})


class MCPRegistry:
    def __init__(self) -> None:
        self._tools: dict[str, MCPTool] = {}

    def register(self, tool: MCPTool) -> None:
        self._tools[tool.name] = tool

    def unregister(self, name: str) -> None:
        self._tools.pop(name, None)

    def list(self) -> list[MCPTool]:
        return list(self._tools.values())


class MCPConfigLoader:
    def load_default(self, cwd: str | None = None) -> MCPConfig:
        root = Path(cwd or os.getcwd())
        config_path = root / ".dispersl" / "mcp.json"
        if not config_path.exists():
            return MCPConfig()
        raw = json.loads(config_path.read_text(encoding="utf-8"))
        return self._interpolate(raw)

    def merge(self, base: MCPConfig, override: dict[str, Any] | None) -> MCPConfig:
        if not override:
            return base
        merged = {
            "version": override.get("version", base.version),
            "servers": {**base.servers, **override.get("servers", {})},
            "tool_policies": {**base.tool_policies, **override.get("tool_policies", {})},
            "defaults": {**base.defaults, **override.get("defaults", {})},
        }
        return MCPConfig(**merged)

    def _interpolate(self, raw: dict[str, Any]) -> MCPConfig:
        text = json.dumps(raw)
        for key, value in os.environ.items():
            text = text.replace(f"${{{key}}}", value)
        return MCPConfig(**json.loads(text))
